contracts/BeakerContract.sol

```diff
@@ -149,6 +149,62 @@ contract BeakerContract is IKernel {
         }
   }
 
+  function proc_acc_call(uint8 capIndex, address account, uint256 amount, bytes memory input) internal returns (uint32 err, bytes memory output) {
+        assembly {
+            function malloc(size) -> result {
+                // align to 32-byte words
+                let rsize := add(size,sub(32,mod(size,32)))
+                // get the current free mem location
+                result :=  mload(0x40)
+                // Bump the value of 0x40 so that it holds the next
+                // available memory location.
+                mstore(0x40,add(result,rsize))
+            }
+
+            let inputSize := mload(input)
+            let bufSize := add(0x80, inputSize)
+
+            let buf := malloc(bufSize)
+
+            // First set up the input data
+            // The acc call call is 0x-09
+            mstore(add(buf,0x0),0x09)
+            // The capability index
+            mstore(add(buf,0x20),capIndex)
+            // The address of the account/contract
+            mstore(add(buf,0x40),account)
+            // The wei to be sent
+            mstore(add(buf,0x60),amount)
+
+            // The data from 0x80 onwards is the data we want to send to
+            // this procedure
+            let inputStart := add(input, 0x20)
+            let bufStart := add(buf, 0x80)
+
+            // An inefficient memcopy
+            for { let n:= 0 } iszero(eq(n, inputSize)) { n := add(n,1)} {
```

Comment:
We should probably discuss this since this is a pretty large expense. Any thoughts in doing a memory swap if `input` location is greater than `0x40 + syscall data size` ?

Reply:
Agreed. At first I wondered why it was written that way, then I realised that because it isn't a CALL we can't use calldata, so it made sense.

I've thought about the same things too. I think we can do it, although obviously we have to be careful because we are playing with some one else's stack, so to speak.

Given the complexity it's something we should profile first I think. How about we make an issue with known optimisation areas, then we can simply insert some log(gas) for simple and precise benchmarking so we can prioritise those.

Reply:
It's probably worth pointing out that this is how BeakerContract.sol was already written. This is just a more uniform implementation.

Reply:
The updated memcopy is in the update-call branch. I've done some calcs and it costs about 3 gas per byte. For reference it costs 68 per byte to send it in the syscall, so it's definitely worth putting on the optimisation list, but it isn't a deal breaker. Much much cheaper than abstracting it away into a library (which would give us that 68 per byte cost twice) which would allows us to take advantage of calldatacopy.

Reply:
To continue this soliloquy, turns out that calldatacopy is 3+3*nbytes, so this is actually cheaper, but the opportunity you mentioned for optimisation still exists.